core/src/memory.rs

```diff
@@ -166,48 +166,62 @@ impl core::ops::Sub for WasmPageNumber {
 /// then we will include this wasm page in result set.
 /// If there is wasm pages, for which `pages_iter` contains not all pages,
 /// then returns Err.
-/// ````
-/// Example: if one wasm page contains 4 gear pages
-/// pages = [0,1,2,3,12,13,14,15]
-/// then result wasm pages = [0,3]
 ///
-/// pages = [0,1,2,3,5,6,7,8]
-/// then result is Err
-/// ````
+/// # Examples
+///
+/// We assume the one wasm page contains 16 gear pages.
+///
+/// ```
+/// # use std::collections::BTreeSet;
+/// # use gear_core::memory::{self, PageNumber, WasmPageNumber};
+///
+/// let gear_pages: BTreeSet<_> = vec![0..16, 48..64].into_iter().flatten().map(PageNumber).collect();
+/// let wasm_pages: BTreeSet<_> = [0, 3].map(WasmPageNumber).into();
+/// assert_eq!(memory::pages_to_wasm_pages_set(gear_pages.iter()), Ok(wasm_pages));
+///
+/// let gear_pages: BTreeSet<_> = vec![0..16, 50..66].into_iter().flatten().map(PageNumber).collect();
+/// assert!(memory::pages_to_wasm_pages_set(gear_pages.iter()).is_err());
+/// ```
 pub fn pages_to_wasm_pages_set<'a>(
-    mut pages_iter: impl Iterator<Item = &'a PageNumber>,
+    pages_iter: impl Iterator<Item = &'a PageNumber>,
 ) -> Result<BTreeSet<WasmPageNumber>, &'static str> {
-    let mut wasm_pages = BTreeSet::<WasmPageNumber>::new();
-    while let Some(page) = pages_iter.next() {
-        if page.0 % PageNumber::num_in_one_wasm_page() != 0 {
-            return Err("There is wasm page, which has not all gear pages in the begin");
-        }
-        let wasm_page_num = WasmPageNumber(page.0 / PageNumber::num_in_one_wasm_page());
-        wasm_pages.insert(wasm_page_num);
-        for _ in 0..(WasmPageNumber::size() / PageNumber::size() - 1) {
-            pages_iter
-                .next()
-                .expect("There is wasm page, which has not all gear pages in the end");
-        }
-    }
+    let mut wasm_pages = BTreeSet::new();
+    pages_iter
```

Comment:
Frankly speaking it looks like there should be guarantee that `pages_iter` is sorted

Reply:
Exactly. I've not changed the function contract.